File: src/main/java/hiiragi283/material/container/ContainerModuleMachine.java
package hiiragi283.material.container;

import hiiragi283.material.api.container.HiiragiContainer;
import hiiragi283.material.api.container.SlotModuleMachine;
import hiiragi283.material.api.container.SlotOutputItemHandler;
import hiiragi283.material.tile.TileEntityModuleMachine;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;
import org.jetbrains.annotations.NotNull;

import javax.annotation.ParametersAreNonnullByDefault;

@ParametersAreNonnullByDefault
public class ContainerModuleMachine extends HiiragiContainer<TileEntityModuleMachine> {

    public ContainerModuleMachine(TileEntityModuleMachine tile, EntityPlayer player) {
        super(tile, player);
        this.addSlotToContainer(new SlotModuleMachine(tile, tile.inventoryInput, 0, 8 + 18, 18));
        this.addSlotToContainer(new SlotModuleMachine(tile, tile.inventoryInput, 1, 8 + 18 * 2, 18));
        this.addSlotToContainer(new SlotModuleMachine(tile, tile.inventoryInput, 2, 8 + 18 * 3, 18));
        this.addSlotToContainer(new SlotModuleMachine(tile, tile.inventoryInput, 3, 8 + 18, 18 * 2));
        this.addSlotToContainer(new SlotModuleMachine(tile, tile.inventoryInput, 4, 8 + 18 * 2, 18 * 2));
        this.addSlotToContainer(new SlotModuleMachine(tile, tile.inventoryInput, 5, 8 + 18 * 3, 18 * 2));
        this.addSlotToContainer(new SlotOutputItemHandler(tile.inventoryOutput, 0, 8 + 18 * 5, 18));
        this.addSlotToContainer(new SlotOutputItemHandler(tile.inventoryOutput, 1, 8 + 18 * 6, 18));
        this.addSlotToContainer(new SlotOutputItemHandler(tile.inventoryOutput, 2, 8 + 18 * 7, 18));
        this.addSlotToContainer(new SlotOutputItemHandler(tile.inventoryOutput, 3, 8 + 18 * 5, 18 * 2));
        this.addSlotToContainer(new SlotOutputItemHandler(tile.inventoryOutput, 4, 8 + 18 * 6, 18 * 2));
        this.addSlotToContainer(new SlotOutputItemHandler(tile.inventoryOutput, 5, 8 + 18 * 7, 18 * 2));
        this.initSlotsPlayer(84);
    }

    @NotNull
    @Override
    public ItemStack transferStackInSlot(EntityPlayer player, int index) {
        ItemStack stack = ItemStack.EMPTY;
        Slot slot = inventorySlots.get(index);
        if (slot.getHasStack()) {
            ItemStack stackSlot = slot.getStack();
            stack = stackSlot.copy();
            switch (index) {
                //Input, Output -> Inventory, Hotbar
                case 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11 -> {
                    if (!mergeItemStack(stackSlot, 4, inventorySlots.size(), false)) return ItemStack.EMPTY;
                }
                //Inventory, Hotbar -> Input
                default -> {
                    if (!mergeItemStack(stackSlot, 0, 4, false)) return ItemStack.EMPTY;
                }
            }
            if (stackSlot.isEmpty()) {
                slot.putStack(ItemStack.EMPTY);
            } else slot.onSlotChanged();
        }
        return stack;
    }

}

File: src/main/kotlin/hiiragi283/material/container/ContainerModuleMachine.kt
package hiiragi283.material.container

import hiiragi283.material.api.container.HiiragiContainer
import hiiragi283.material.api.container.SlotItemHandlerControllable
import hiiragi283.material.network.HiiragiMessage
import hiiragi283.material.network.HiiragiNetworkWrapper
import hiiragi283.material.tile.TileEntityModuleMachine
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.inventory.Slot
import net.minecraft.item.ItemStack
import net.minecraftforge.items.IItemHandler
import net.minecraftforge.items.SlotItemHandler


class ContainerModuleMachine(tile: TileEntityModuleMachine, player: EntityPlayer) :
    HiiragiContainer.TileEntity<TileEntityModuleMachine>(tile, player) {

    init {
        addSlotToContainer(SlotModuleMachine(tile, tile.inventoryInput, 0, getSlotPosX(1), getSlotPosY(0)))
        addSlotToContainer(SlotModuleMachine(tile, tile.inventoryInput, 1, getSlotPosX(2), getSlotPosY(0)))
        addSlotToContainer(SlotModuleMachine(tile, tile.inventoryInput, 2, getSlotPosX(3), getSlotPosY(0)))
        addSlotToContainer(SlotModuleMachine(tile, tile.inventoryInput, 3, getSlotPosX(1), getSlotPosY(1)))
        addSlotToContainer(SlotModuleMachine(tile, tile.inventoryInput, 4, getSlotPosX(2), getSlotPosY(1)))
        addSlotToContainer(SlotModuleMachine(tile, tile.inventoryInput, 5, getSlotPosX(3), getSlotPosY(1)))
        addSlotToContainer(SlotItemHandlerControllable(tile.inventoryOutput, 0, getSlotPosX(5), getSlotPosY(0)))
        addSlotToContainer(SlotItemHandlerControllable(tile.inventoryOutput, 1, getSlotPosX(6), getSlotPosY(0)))
        addSlotToContainer(SlotItemHandlerControllable(tile.inventoryOutput, 2, getSlotPosX(7), getSlotPosY(0)))
        addSlotToContainer(SlotItemHandlerControllable(tile.inventoryOutput, 3, getSlotPosX(5), getSlotPosY(1)))
        addSlotToContainer(SlotItemHandlerControllable(tile.inventoryOutput, 4, getSlotPosX(6), getSlotPosY(1)))
        addSlotToContainer(SlotItemHandlerControllable(tile.inventoryOutput, 5, getSlotPosX(7), getSlotPosY(1)))
        initSlotsPlayer(84)
        HiiragiNetworkWrapper.sendToAll(HiiragiMessage.SyncCount(tile.pos, tile.currentCount))
    }

    override fun transferStackInSlot(player: EntityPlayer, index: Int): ItemStack {
        var stack: ItemStack = ItemStack.EMPTY
        val slot: Slot = inventorySlots[index]
        if (slot.hasStack) {
            val stackSlot: ItemStack = slot.stack
            stack = stackSlot.copy()
            if (index in (0..11)) {
                if (!mergeItemStack(stackSlot, 12, inventorySlots.size, true)) return ItemStack.EMPTY
            } else {
                if (!mergeItemStack(stackSlot, 0, 6, false )) return ItemStack.EMPTY
            }
            if (stackSlot.isEmpty) {
                slot.putStack(ItemStack.EMPTY)
            } else slot.onSlotChanged()
        }
        return stack
    }

    class SlotModuleMachine(
        private val tile: TileEntityModuleMachine,
        itemHandler: IItemHandler,
        index: Int,
        xPosition: Int,
        yPosition: Int
    ) : SlotItemHandler(itemHandler, index, xPosition, yPosition) {

        override fun isItemValid(stack: ItemStack): Boolean = tile.machineProperty.itemSlots >= slotIndex + 1

        override fun isEnabled(): Boolean = tile.machineProperty.itemSlots >= slotIndex + 1

    }

}